### PR TITLE
Support quantized dot merge (share operand)

### PR DESCRIPTION
**Pull request by linbinskn**

### CODE REVIEW

File: pytorch_blade/tests/disc/pdl/test_e2e/test_quantized_qkv_merge.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# Copyright 2021 The BladeDISC Authors. All rights reserved.
```
Comment (deeply) [inline pytorch_blade/tests/disc/pdl/test_e2e/test_quantized_qkv_merge.py:1]: welcome to 2023 😂

File: tao_compiler/mlir/disc/disc_compiler.cc [hunk 2]
```diff
@@ -276,6 +276,14 @@ LogicalResult LowerHLOToLLVM(ModuleOp m, const DISCLoweringOptions& options) {
     pm.addNestedPass<FuncOp>(disc_ral::createDiscDotMergePass());
   }
 
+  bool enable_quantized_dot_merge = false;
+  tensorflow::ReadBoolFromEnvVar("BLADE_GEMM_TUNE_JIT",
+                                 enable_quantized_dot_merge,
+                                 &enable_quantized_dot_merge);
+  if (enable_quantized_dot_merge == true) {
```
Comment (deeply) [inline tao_compiler/mlir/disc/disc_compiler.cc:283]: nitpicking: `if (enable_quantized_dot_merge) {`

File: tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc [hunk 3]
```diff
@@ -183,6 +183,143 @@ struct GemmScaleState : public Context::Resource {
...
+  {
+    float alpha = 1.0f;
+    beta = 0.0f;
+    void* s = gpu_driver->asCUStream(ctx, stream_handle);
+    bladnn::Context bladnn_ctx{s};
+    bladnn::Dtype in_dtype = bladnn::Dtype::kS8;
+    bladnn::Dtype out_dtype = bladnn::Dtype::kS8;
+    bool ret = false;
+
+    ret = bladnn::gemm(&bladnn_ctx, in_dtype, 0, input.data, m, k, in_dtype, 1,
```
Comment (deeply) [inline tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc:308]: nitpicking: `bool ret = bladnn::gemm(...)`

File: pytorch_blade/tests/disc/pdl/test_e2e/test_quantized_qkv_merge.py (hunk 1 above)
Comment (linbinskn) [reply pytorch_blade/tests/disc/pdl/test_e2e/test_quantized_qkv_merge.py]: done ovo

File: tao_compiler/mlir/disc/disc_compiler.cc (hunk 2 above)
Comment (linbinskn) [reply tao_compiler/mlir/disc/disc_compiler.cc]: done

File: tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc (hunk 3 above)
Comment (linbinskn) [reply tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc]: done

File: tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc [hunk 4]
```diff
@@ -183,6 +183,143 @@ struct GemmScaleState : public Context::Resource {
...
 
+struct GemmScaleChannelState : public Context::Resource {
+  std::mutex mu;
+  std::unordered_map<GemmScaleKey, std::vector<float*>, GemmScaleKeyHasher>
+      cache;
+};
+
+// Layout requirement:
+// input: m x k, row major
+// weight: n x k, col major
```
Comment (wyzero) [inline tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc:194]: the shape of weight is already `nxk`, so should be row major?

File: tao_compiler/mlir/disc/transforms/disc_quantized_dot_merge.cc [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+#include "mlir/Transforms/Passes.h"
+#include "tensorflow/compiler/mlir/disc/IR/hlo_disc_ops.h"
+#include "tensorflow/compiler/mlir/disc/disc_util.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+#include "tensorflow/compiler/mlir/disc/transforms/shape_utils.h"
+#include "tensorflow/compiler/mlir/disc/utils/cycle_detector.h"
+
+#define DEBUG_TYPE "disc-quantized-dot-merge"
+
+namespace mlir {
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/disc_quantized_dot_merge.cc:32]: is there any code can be shared with normal dot merge pass?  it seems there is a lot of duplicated code.

File: tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc (hunk 4 above)
Comment (linbinskn) [reply tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc:194]: for weight, `kxn` is row major and `nxk` is colmajor?

File: tao_compiler/mlir/disc/transforms/disc_quantized_dot_merge.cc (hunk 5 above)
Comment (linbinskn) [reply tao_compiler/mlir/disc/transforms/disc_quantized_dot_merge.cc:32]: Most of code differs from dot merge pass. May be `TryMergeQuantizedDotClusters` and `BuildQuantizedDotClusters` can be shared with it. I will try to merge them.

File: tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc (hunk 4 above)
Comment (wyzero) [reply tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc:194]: I see. Maybe just different convention:
`memref<kxn>, row major, strides = [n, 1]`
`memref<kxn>, column major, strides = [1, k]`
`memref<nxk>, column major, strides = [1, n]`
`memref<nxk>, row major, strides = [k, 1]`

**Review by wyzero — APPROVED** [review]

**Pull request merged**